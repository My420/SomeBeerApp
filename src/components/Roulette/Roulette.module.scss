@import '../../sass/variables.scss';

$cell-width: 120px;
$cell-height: 120px;

$screen-border-width: 6px;
$cell-border-width: 2px;

$row-cell-amount: 6;
$window-cell-amount: 4;

$pointer-width: $cell-border-width * 2;

.wrapper {
  width: ($cell-width * $window-cell-amount) + ($screen-border-width * 2);
  margin: 0 auto;

  outline: 2px solid green;
}

.window {
  display: flex;
  position: relative;

  width: $cell-width * $window-cell-amount;
  height: $cell-height;

  border: $screen-border-width solid $menu-bar-color;

  overflow: hidden;
  list-style: none;
}

.pointer {
  position: absolute;
  top: 0px;
  left: ($cell-width * $window-cell-amount / 2) - $pointer-width / 2;
  z-index: 3;

  width: $pointer-width;
  height: 100%;

  background-color: yellowgreen;
}

.firstRow,
.secondRow {
  display: flex;

  width: $cell-width * $row-cell-amount;
  height: $cell-height;

  list-style: none;
}

.cell {
  box-sizing: border-box;

  width: $cell-width;
  height: $cell-height;

  border: $cell-border-width solid $menu-hover-color;

  overflow: hidden;
}

.startButton {
  display: block;
  margin: 30px auto;
}

@media (max-width: 550px) {
  $cell-width: 86px;
  $cell-height: 86px;

  .wrapper {
    width: ($cell-width * $window-cell-amount) + ($screen-border-width * 2);
  }

  .window {
    width: $cell-width * $window-cell-amount;
    height: $cell-height;
  }

  .pointer {
    left: ($cell-width * $window-cell-amount / 2) - $pointer-width / 2;
  }

  .firstRow,
  .secondRow {
    width: $cell-width * $row-cell-amount;
    height: $cell-height;
  }

  .cell {
    width: $cell-width;
    height: $cell-height;
  }

  .startButton {
    display: block;
    margin: 30px auto;
  }
}

@media (max-width: 400px) {
  $cell-width: 60px;
  $cell-height: 60px;

  $screen-border-width: 4px;

  .wrapper {
    width: ($cell-width * $window-cell-amount) + ($screen-border-width * 2);
  }

  .window {
    width: $cell-width * $window-cell-amount;
    height: $cell-height;

    border: $screen-border-width solid $menu-bar-color;
  }

  .pointer {
    left: ($cell-width * $window-cell-amount / 2) - $pointer-width / 2;
  }

  .firstRow,
  .secondRow {
    width: $cell-width * $row-cell-amount;
    height: $cell-height;
  }

  .cell {
    width: $cell-width;
    height: $cell-height;
  }

  .startButton {
    display: block;
    margin: 30px auto;
  }
}
