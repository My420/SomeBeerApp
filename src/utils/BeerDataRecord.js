import { Record } from 'immutable';

const BeerDataRecord = Record(
  {
    id: null,
    name: null,
    tagline: null,
    first_brewed: null,
    description: null,
    image_url: null,
    abv: null,
    ibu: null,
    target_fg: null,
    target_og: null,
    ebc: null,
    srm: null,
    ph: null,
    attenuation_level: null,
    volume: {},
    boil_volume: {},
    method: {},
    ingredients: {},
    food_pairing: [],
    brewers_tips: null,
    contributed_by: null
  },
  'BeerDataRecord'
);

export default BeerDataRecord;
