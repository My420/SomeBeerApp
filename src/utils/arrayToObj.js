/* eslint no-param-reassign: "error" */

const arrayToObj = arr => {
  return arr.reduce((prevValue, element) => {
    prevValue[element.id] = element;
    return prevValue;
  }, {});
};

export default arrayToObj;
