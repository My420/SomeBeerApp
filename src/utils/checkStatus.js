const checkStatus = response => {
  const responseStatus = +response.status;
  if (responseStatus >= 200 && responseStatus < 300) {
    return response;
  }
  throw new Error(response.statusText);
};

export default checkStatus;
