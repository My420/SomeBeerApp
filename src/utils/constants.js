export const SHOP_NAME = 'Some Beer Shop';

export const LINK = {
  HOME: '/',
  CATALOG: '/catalog',
  ROULETTE: '/roulette',
  ABOUT: '/about',
  FAVORITE: '/favorite',
  CART: '/cart'
};

export const NAVIGATION_LINKS = [
  [LINK.HOME, 'Home'],
  [LINK.CATALOG, 'Catalog'],
  [LINK.ROULETTE, 'Roulette!'],
  [LINK.FAVORITE, 'Favorite'],
  [LINK.ABOUT, 'About'],
  [LINK.CART, 'Cart']
];

export const BROWSER_WIDTH_RATIO = {
  LPC: 4,
  MPC: 3,
  TABLET_9: 2,
  TABLET_7: 1,
  MOBILE: 0
};

export const BROWSER_MIN_WIDTH = {
  LPC: 1200,
  MPC: 993,
  TABLET_9: 768,
  TABLET_7: 480,
  MOBILE: 320
};

export const SOCIAL_LINK = {
  GITHUB: 'https://github.com/My420',
  EMAIL: 'mailto:klinovitsky.aleksey@gmail.com',
  SKYPE: 'skype:live:klinovitsky.aleksey?chat'
};

export const AUTHOR = 'Klinovitsky Aleksey';

export const POPULAR_BEER__GET_REQUEST = 'POPULAR_BEER__GET_REQUEST';
export const POPULAR_BEER__GET_SUCCESS = 'POPULAR_BEER__GET_SUCCESS';
export const POPULAR_BEER__GET_FAILURE = 'POPULAR_BEER__GET_FAILURE';

export const API_HOST = 'https://api.punkapi.com/v2/';

export const POPULAR_BEER_AMOUNT = 8;

export const LOAD_ERROR_MESSAGE =
  'Could not download data, please refresh the page';

export const ERROR_TITLE = 'Oops! Something went wrong.';
export const ERROR_ACTION = 'Please refresh page or try again later.';
