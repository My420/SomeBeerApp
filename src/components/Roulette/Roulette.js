import React from 'react';
import PropTypes from 'prop-types';
import styles from './Roulette.module.scss';
import ListGroup from '../ListGroup/ListGroup';
import ListGroupItem from '../ListGroupItem/ListGroupItem';
import { ROULETTE_CELLS_AMOUNT } from '../../utils/constants';
import Button from '../Button/Button';
import generateID from '../../utils/generateID';
import spin from '../../utils/spin';

const Roulette = ({ components }) => {
  // eslint-disable-next-line no-console
  console.log('render ==================== Roulette');

  const firstRowRef = React.createRef();
  const secondRowRef = React.createRef();
  const middle = ROULETTE_CELLS_AMOUNT / 2;

  const cells = components.map(elem => {
    return (
      <ListGroupItem className={styles.cell} key={generateID()}>
        {elem}
      </ListGroupItem>
    );
  });

  const onButtonClick = async () => {
    const firstRow = firstRowRef.current;
    const secondRow = secondRowRef.current;
    spin(firstRow, secondRow);
  };

  return (
    <div className={styles.wrapper}>
      <ListGroup className={styles.window}>
        <div className={styles.pointer} />
        <div className={styles.firstRow} ref={firstRowRef}>
          {cells.slice(0, middle)}
        </div>
        <div className={styles.secondRow} ref={secondRowRef}>
          {cells.slice(middle, ROULETTE_CELLS_AMOUNT + 1)}
        </div>
      </ListGroup>
      <Button className={styles.startButton} onClick={onButtonClick}>
        Start
      </Button>
    </div>
  );
};

Roulette.propTypes = {
  components: PropTypes.arrayOf(PropTypes.node).isRequired
};

export default Roulette;
