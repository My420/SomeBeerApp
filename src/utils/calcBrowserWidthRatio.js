import { BROWSER_MIN_WIDTH, BROWSER_WIDTH_RATIO } from './constants';

const calcBrowserWidthRatio = width => {
  if (width > BROWSER_MIN_WIDTH.LPC) {
    return BROWSER_WIDTH_RATIO.LPC;
  }
  if (width > BROWSER_MIN_WIDTH.MPC) {
    return BROWSER_WIDTH_RATIO.MPC;
  }
  if (width > BROWSER_MIN_WIDTH.TABLET) {
    return BROWSER_WIDTH_RATIO.TABLET_9;
  }
  if (width > BROWSER_MIN_WIDTH.MOBILE) {
    return BROWSER_WIDTH_RATIO.TABLET_7;
  }
  return BROWSER_WIDTH_RATIO.MOBILE;
};

export default calcBrowserWidthRatio;
