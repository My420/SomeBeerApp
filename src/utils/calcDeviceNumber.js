import { DEVICE_MIN_WIDTH, DEVICE_NUMBER } from './constants';

const calcDeviceNumber = width => {
  if (width > DEVICE_MIN_WIDTH.LPC) {
    return DEVICE_NUMBER.LPC;
  }
  if (width > DEVICE_MIN_WIDTH.MPC) {
    return DEVICE_NUMBER.MPC;
  }
  if (width > DEVICE_MIN_WIDTH.TABLET) {
    return DEVICE_NUMBER.TABLET;
  }
  return DEVICE_NUMBER.MOBILE;
};

export default calcDeviceNumber;
