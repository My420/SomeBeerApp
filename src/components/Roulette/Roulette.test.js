import React from 'react';
import { shallow } from 'enzyme';
import Roulette from './Roulette';
import { ROULETTE_CELLS_AMOUNT } from '../../utils/constants';

describe('Test <Roulette/> component', () => {
  const TestComp = id => {
    return <p>{id}</p>;
  };

  const components = [];
  for (let i = 0; i < ROULETTE_CELLS_AMOUNT; i += 1) {
    components.push(<TestComp id={i + 1} />);
  }

  const firstRowX = 20;
  const secondRowX = 400;
  const changeMock = jest.fn();
  const winnerMock = jest.fn();

  const wrapper = shallow(
    <Roulette
      components={components}
      firstRowX={firstRowX}
      secondRowX={secondRowX}
      changeData={changeMock}
      onNewWinner={winnerMock}
    />
  );

  test('should correct render component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render ListGroup component', () => {
    const component = wrapper.find('ListGroup');
    expect(component).toHaveLength(1);
  });

  test('should render tag <div/> with className = pointer inside ListGroup component', () => {
    const component = wrapper.find('ListGroup').find('div.pointer');
    expect(component).toHaveLength(1);
  });

  test('should render tag <div/> with className = firstRow inside ListGroup component', () => {
    const component = wrapper.find('ListGroup').find('div.firstRow');
    expect(component).toHaveLength(1);
  });

  test('should render tag <div/> with className = secondRow inside ListGroup component', () => {
    const component = wrapper.find('ListGroup').find('div.secondRow');
    expect(component).toHaveLength(1);
  });

  test('tag <div/> with className = firstRow should have correct style attr', () => {
    const component = wrapper.find('ListGroup').find('div.firstRow');
    expect(component.prop('style').transform).toBe(
      `translateX(${firstRowX}px)`
    );
  });

  test('tag <div/> with className = secondRow should have correct style attr', () => {
    const component = wrapper.find('ListGroup').find('div.secondRow');
    expect(component.prop('style').transform).toBe(
      `translateX(${secondRowX}px)`
    );
  });

  test(`should render ${ROULETTE_CELLS_AMOUNT /
    2} ListGroupItem component inside tag <div/> with className = secondRow`, () => {
    const component = wrapper.find('div.secondRow').find('ListGroupItem');
    expect(component).toHaveLength(6);
  });

  test(`should render ${ROULETTE_CELLS_AMOUNT /
    2} ListGroupItem component inside tag <div/> with className = firstRow`, () => {
    const component = wrapper.find('div.firstRow').find('ListGroupItem');
    expect(component).toHaveLength(6);
  });

  test('inside ListGroupItem component should be correct TestComp component', () => {
    const component = wrapper.find('ListGroupItem');
    component.forEach((elem, i) => {
      const expectedComp = elem.find('TestComp');
      expect(expectedComp).toHaveLength(1);
      expect(expectedComp.prop('id')).toBe(i + 1);
    });
  });

  test('should correct render change button', () => {
    const component = wrapper.find('button.changeButton');
    expect(component).toHaveLength(1);
    expect(component.text()).toBe('Change');
  });

  test("click on change button should call 'changeData' prop func", () => {
    const component = wrapper.find('button.changeButton');
    component.simulate('click');
    expect(changeMock).toHaveBeenCalledTimes(1);
  });

  test('should correct render spin button', () => {
    const component = wrapper.find('button.startButton');
    expect(component).toHaveLength(1);
    expect(component.text()).toBe('Spin');
  });
});
