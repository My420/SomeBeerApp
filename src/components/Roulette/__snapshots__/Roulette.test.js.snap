// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Roulette/> component should correct render component 1`] = `
<div
  className="wrapper"
>
  <ListGroup
    className="window"
    tag="ul"
  >
    <div
      className="pointer"
    />
    <div
      className="firstRow"
      style={
        Object {
          "transform": "translateX(20px)",
        }
      }
    >
      <ListGroupItem
        className="cell"
        key="1"
      >
        <TestComp
          id={1}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="2"
      >
        <TestComp
          id={2}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="3"
      >
        <TestComp
          id={3}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="4"
      >
        <TestComp
          id={4}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="5"
      >
        <TestComp
          id={5}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="6"
      >
        <TestComp
          id={6}
        />
      </ListGroupItem>
    </div>
    <div
      className="secondRow"
      style={
        Object {
          "transform": "translateX(400px)",
        }
      }
    >
      <ListGroupItem
        className="cell"
        key="7"
      >
        <TestComp
          id={7}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="8"
      >
        <TestComp
          id={8}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="9"
      >
        <TestComp
          id={9}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="10"
      >
        <TestComp
          id={10}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="11"
      >
        <TestComp
          id={11}
        />
      </ListGroupItem>
      <ListGroupItem
        className="cell"
        key="12"
      >
        <TestComp
          id={12}
        />
      </ListGroupItem>
    </div>
  </ListGroup>
  <div
    className="controls"
  >
    <button
      className="startButton"
      onClick={[Function]}
      type="button"
    >
      Spin
    </button>
    <button
      className="changeButton"
      onClick={[MockFunction]}
      type="button"
    >
      Change
    </button>
  </div>
</div>
`;
